src/PositionDescriptor.sol

```diff
@@ -0,0 +1,132 @@
+import {IPositionManager} from "./interfaces/IPositionManager.sol";
+import {IPositionDescriptor} from "./interfaces/IPositionDescriptor.sol";
+import {PositionInfo, PositionInfoLibrary} from "./libraries/PositionInfoLibrary.sol";
+import {Descriptor} from "./libraries/Descriptor.sol";
+import {CurrencyRatioSortOrder} from "./libraries/CurrencyRatioSortOrder.sol";
+import {SafeERC20Namer} from "./libraries/SafeERC20Namer.sol";
+
+/// @title Describes NFT token positions
+/// @notice Produces a string containing the data URI for a JSON metadata string
+contract PositionDescriptor is IPositionDescriptor {
+    using StateLibrary for IPoolManager;
+    using PoolIdLibrary for PoolKey;
+    using CurrencyLibrary for Currency;
+    using PositionInfoLibrary for PositionInfo;
+
+    // mainnet addresses
+    address private constant DAI = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address private constant USDC = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address private constant USDT = 0xdAC17F958D2ee523a2206206994597C13D831ec7;
+    address private constant TBTC = 0x8dAEBADE922dF735c38C80C7eBD708Af50815fAa;
+    address private constant WBTC = 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599;
+
+    address public immutable WETH9;
+    string public nativeCurrencyLabel;
+
+    IPoolManager public immutable poolManager;
+
+    constructor(IPoolManager _poolManager, address _WETH9, string memory _nativeCurrencyLabel) {
+        poolManager = _poolManager;
+        WETH9 = _WETH9;
+        nativeCurrencyLabel = _nativeCurrencyLabel;
+    }
+
+    /// @inheritdoc IPositionDescriptor
+    function tokenURI(IPositionManager positionManager, uint256 tokenId)
+        external
+        view
+        override
+        returns (string memory)
+    {
+        (PoolKey memory poolKey, PositionInfo positionInfo) = positionManager.getPoolAndPositionInfo(tokenId);
+        (, int24 tick,,) = poolManager.getSlot0(poolKey.toId());
+
+        // If possible, flip currencies to get the larger currency as the base currency, so that the price (quote/base) is more readable
+        // flip if currency0 priority is greater than currency1 priority
+        bool _flipRatio = flipRatio(Currency.unwrap(poolKey.currency0), Currency.unwrap(poolKey.currency1));
+
+        // If not flipped, quote currency is currency1, base currency is currency0
+        // If flipped, quote currency is currency0, base currency is currency1
+        Currency quoteCurrency = !_flipRatio ? poolKey.currency1 : poolKey.currency0;
+        Currency baseCurrency = !_flipRatio ? poolKey.currency0 : poolKey.currency1;
+
+        return Descriptor.constructTokenURI(
+            Descriptor.ConstructTokenURIParams({
+                tokenId: tokenId,
+                quoteCurrency: quoteCurrency,
+                baseCurrency: baseCurrency,
+                quoteCurrencySymbol: quoteCurrency.isAddressZero()
```

Comment:
i think id rather this logic was inside the namer. Like SafeCurrencyNamer and then in there check `.isAddressZero()` to abstract away from this contract

Reply:
so have the isAddressZero check inside of the namer? that's not directly related to name though. Plus, the isAddressZero check is also used for decimals

Reply:
I guess I'm wondering if `SafeERC20Namer` should be called like `SafeCurrencyMetadata` or something?
and then it can include
- logic for checking for isAddressZero and returning information for native
- try-catching decimals (from other comment) and returning 0 if not